### PR TITLE
[Fancode] Add new extractor

### PR DESCRIPTION
**Pull request by rhsmachine**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

New extractor for fancode, an Indian sports streaming site. This PR adds support for regular videos (not live). https://github.com/yt-dlp/yt-dlp/issues/269

### CODE REVIEW

File: yt_dlp/extractor/fancode.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                "id":%s,
+                "filter":{
+                    "contentDataType":"DEFAULT"
+                }
+            },
+            "operationName":"Video"
+            }'''
+
+        metadata_json = self._download_json(GQL_URL, video_id, data=str.encode(data % video_id), headers=headers, note="Downloading metadata")
+        brightcove_video_id = metadata_json.get('data').get('media').get('mediaSource').get('brightcove')
```
Comment (Ashish0804) [inline yt_dlp/extractor/fancode.py:66]: Do not use `.get` chains.
Use `try_get` from `utils`
```python
brightcove_video_id = try_get(metadata_json, lambda x: x['data']['media']['mediaSource']['brightcove'], compat_str)
```

File: yt_dlp/extractor/fancode.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        if brightcove_video_id is None:
+            raise RegexNotFoundError('Unable to extract brightcove Video ID')
+
+        if metadata_json.get('data').get('media').get('isPremium') is True:
```
Comment (Ashish0804) [inline yt_dlp/extractor/fancode.py:71]: Same as above.

File: yt_dlp/extractor/fancode.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return {
+            '_type': 'url_transparent',
+            'url': BRIGHTCOVE_URL_TEMPLATE % (brightcove_user_id, brightcove_video_id),
+            'ie_key': 'BrightcoveNew',
+            'id': video_id,
+            'title': metadata_json.get('data').get('media').get('title'),
+            'like_count': metadata_json.get('data').get('media').get('totalUpvotes'),
+            'view_count': metadata_json.get('data').get('media').get('totalViews'),
+            'tags': metadata_json.get('data').get('media').get('tags'),
+            'release_timestamp': parse_iso8601(metadata_json.get('data').get('media').get('publishedTime'))
```
Comment (Ashish0804) [inline yt_dlp/extractor/fancode.py:83]: Same as above.

File: yt_dlp/extractor/fancode.py [hunk 4]
```diff
@@ -63,22 +66,24 @@ def _real_extract(self, url):
...
+        brightcove_video_id = try_get(metadata_json, lambda x: x['data']['media']['mediaSource']['brightcove'], compat_str)
 
         if brightcove_video_id is None:
             raise RegexNotFoundError('Unable to extract brightcove Video ID')
 
-        if metadata_json.get('data').get('media').get('isPremium') is True:
+        if try_get(metadata_json, lambda x: x['data']['media']['isPremium']) is True:
             self.report_warning("requires a premium account" % video_id, video_id)
 
+        media = try_get(metadata_json, lambda x: x['data']['media'])
```
Comment (Ashish0804) [inline yt_dlp/extractor/fancode.py]: You can move `media` above the premium check, no?
then you can do `media.get('isPremium')`

File: yt_dlp/extractor/fancode.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                }
+            },
+            "operationName":"Video"
+            }'''
+
+        metadata_json = self._download_json(GQL_URL, video_id, data=str.encode(data % video_id), headers=headers, note="Downloading metadata")
+        brightcove_video_id = try_get(metadata_json, lambda x: x['data']['media']['mediaSource']['brightcove'], compat_str)
+
+        if brightcove_video_id is None:
+            raise RegexNotFoundError('Unable to extract brightcove Video ID')
```
Comment (pukkandan) [inline yt_dlp/extractor/fancode.py:72]:
```suggestion
            raise ExtractorError('Unable to extract brightcove Video ID')
```
This is definitely not a `RegexNotFoundError`

**Review by pukkandan — COMMENTED** [review]
Some comments in addition to @Ashish0804's review

Also, use `'` everywhere for consistency, not `"`

File: yt_dlp/extractor/fancode.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        BRIGHTCOVE_URL_TEMPLATE = 'https://players.brightcove.net/%s/default_default/index.html?videoId=%s'
+
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        brightcove_user_id = self._html_search_regex(
+            r'(?:https?://)?players\.brightcove\.net/(\d+)/default_default/index(?:\.min)?\.js',
+            webpage, "user id")
+
+        headers = {
+            'content-type': 'application/json',  # has to be set otherwise api errors out
```
Comment (pukkandan) [inline yt_dlp/extractor/fancode.py:49]:
```suggestion
            'content-type': 'application/json',
```

**Review by pukkandan — APPROVED** [review]

**Comment by pukkandan** [comment]
Already merged separately. See #269 for details

**Pull request closed without merging**